nexus/db-queries/src/db/pub_test_utils/mod.rs

```diff
@@ -0,0 +1,316 @@
+            }
+            (Populate::SchemaAndData, Interface::Datastore) => {
+                let db = crdb::test_setup_database(log).await;
+                let (opctx, datastore) =
+                    datastore_test_on_default_rack(log, &db).await;
+                TestDatabase {
+                    db,
+                    kind: TestKind::Datastore { opctx, datastore },
+                }
+            }
+            (Populate::SchemaAndData, Interface::Nothing)
+            | (Populate::SchemaAndData, Interface::Pool) => {
+                // This configuration isn't wrong, it's just weird - we need to
+                // build a datastore to load the built-in data, so it's odd to
+                // discard it immediately afterwards.
+                panic!("If you're fully populating a datastore, you probably want a connection to it");
+            }
+        }
+    }
+}
+
+enum TestKind {
+    NoPool,
+    Pool { pool: Arc<db::Pool> },
+    RawDatastore { datastore: Arc<DataStore> },
+    Datastore { opctx: OpContext, datastore: Arc<DataStore> },
+}
+
+/// A test database, possibly with a pool or full datastore on top
+pub struct TestDatabase {
+    db: CockroachInstance,
+    kind: TestKind,
+}
+
+impl TestDatabase {
+    /// Creates a new database for test usage, without any schema nor interface
+    ///
+    /// [`Self::terminate`] should be called before the test finishes.
+    pub async fn new_without_schema(log: &Logger) -> Self {
+        TestDatabaseBuilder::new()
+            .without_schema()
+            .no_interface()
+            .build(log)
+            .await
+    }
+
+    /// Creates a new database for test usage, with a schema but no interface
+    ///
+    /// [`Self::terminate`] should be called before the test finishes.
+    pub async fn new_with_schema_only(log: &Logger) -> Self {
+        TestDatabaseBuilder::new().with_schema().no_interface().build(log).await
+    }
+
+    /// Creates a new database for test usage, with a pool.
+    ///
+    /// [`Self::terminate`] should be called before the test finishes.
+    pub async fn new_with_pool(log: &Logger) -> Self {
+        TestDatabaseBuilder::new().with_schema().with_pool().build(log).await
+    }
+
+    /// Creates a new database for test usage, with a pre-loaded datastore.
+    ///
+    /// [`Self::terminate`] should be called before the test finishes.
+    pub async fn new_with_datastore(log: &Logger) -> Self {
+        TestDatabaseBuilder::new()
+            .with_schema_and_builtin_data()
+            .with_datastore()
+            .build(log)
+            .await
+    }
+
+    /// Creates a new database for test usage, with a schema but no builtin data
+    ///
+    /// [`Self::terminate`] should be called before the test finishes.
+    pub async fn new_with_raw_datastore(log: &Logger) -> Self {
+        TestDatabaseBuilder::new()
+            .with_schema()
+            .with_datastore()
+            .build(log)
+            .await
+    }
```

Comment:
Given that the builder is also public, do we really need this big pile of constructors? Or, should the builder be a private implementation detail, if we feel it's nicer for users to call these constructors?

Reply:
I'm split on this! I think the builder API is much more explicit, but it's wordier. On the other hand, this `TestDatabase` API is more brief, but slightly less clear.

I exposed both, but admittedly, no one is using the builder API right now. I could make it private until there's a desire to use it?

Reply:
I went ahead and made the builder private in f78d2bb4dcafd59540d23b57f0d3a8e27efb3317